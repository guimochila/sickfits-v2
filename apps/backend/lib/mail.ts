import nodemailer from 'nodemailer'

class MailService {
  static templateEmail(text: string): string {
    return `
     <div styles="
      border: 1px solid black;
      padding: 20px;
      font-family: sans-serif;
      line-height: 2;
      font-size: 20px;
     ">
      <h2>Hello there,</h2>
      <p>${text}</p>
      <p>üëç, Sick Fits team </p>
    </div>
    `
  }

  static async sendMail(resetToken: string, to: string) {
    const transporter = nodemailer.createTransport({
      host: process.env.MAIL_HOST,
      port: parseInt(process.env.MAIL_PORT),
      auth: {
        user: process.env.MAIL_USER,
        pass: process.env.MAIL_PASS,
      },
    })

    await transporter.sendMail({
      to,
      from: 'team@sickfits.com',
      subject: 'Your password reset token',
      html: this.templateEmail(`Your password reset token is here! 
      
      <a href="${process.env.FRONTEND_URL}/reset?token=${resetToken}">Click Here to reset</a>
      `),
    }, (error, info) => {

      console.log(nodemailer.getTestMessageUrl(info))
    })

  }
}

export default MailService
