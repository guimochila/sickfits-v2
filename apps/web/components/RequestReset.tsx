import { useRouter } from 'next/router'
import { FormEvent, useState } from 'react'
import useForm from '../hooks/useForm'
import useRequestReset from '../hooks/useRequestReset'
import DisplayError from './DisplayError'
import Form from './styled/Form'

function RequestReset() {
  const [error, setError] = useState<Record<'message', string> | null>(null)
  const router = useRouter()
  const { inputs, handleChange, resetForm } = useForm({
    email: '',
  })
  const resetMutation = useRequestReset(inputs.email)

  const handleSubmit = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault()
    resetMutation.mutate()
  }

  return (
    <Form method="post" onSubmit={handleSubmit}>
      <DisplayError error={error} />
      {resetMutation.data?.sendUserPasswordResetLink ? <p>Success!</p> : null}
      <h2>Request Password reset</h2>
      <fieldset>
        <label htmlFor="email">
          Email
          <input
            type="email"
            name="email"
            placeholder="Your email address"
            autoComplete="email"
            value={inputs.email}
            onChange={handleChange}
          />
        </label>

        <button type="submit">Request Reset</button>
      </fieldset>
    </Form>
  )
}

export default RequestReset
