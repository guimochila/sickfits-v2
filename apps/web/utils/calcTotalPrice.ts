import { CartItem } from '../@types/graphql-generated'

function calcTotalPrice(cart: CartItem[]) {
  return cart.reduce((total, cartItem) => {
    if (!cartItem.product) return total

    return total + cartItem.quantity * cartItem.product.price
  }, 0)
}

export default calcTotalPrice
