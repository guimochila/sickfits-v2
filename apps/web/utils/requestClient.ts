import { GraphQLClient } from 'graphql-request'
import { API_ENDPOINT } from '../config'

const gqlClient = new GraphQLClient(API_ENDPOINT, {
  credentials: 'include',
  mode: 'cors',
})

export default gqlClient
