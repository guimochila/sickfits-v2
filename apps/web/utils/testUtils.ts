import casual from 'casual'
import { PAGINATION_QUERY } from '../hooks/useProductsCount'
// seed it so we get consistent results
casual.seed(777)

const fakeItem = (): any => ({
  // __typename: 'Product',
  id: 'abc123',
  price: 5000,
  user: null,
  photo: {
    id: 'abc123',
    altText: 'dogs are best',
    image: {
      publicUrlTransformed: 'http://localhost/dog.jpg',
      id: '5e2a13ff689b2835ae71d1a7',
      filename: 'dog.jpg',
      originalFilename: 'dog.jpg',
      mimetype: 'image/jpeg',
      encoding: '7bit',
      publicUrl: 'http://localhost/dog.jpg',
    },
  },
  name: 'dogs are best',
  description: 'dogs',
  status: 'AVAILABLE',
})

const fakeUser = (overrides = {}) => ({
  __typename: 'User',
  id: '4234',
  name: casual.name,
  email: casual.email,
  permissions: ['ADMIN'],
  orders: [],
  cart: [],
  ...overrides,
})

const fakeOrderItem = () => ({
  __typename: 'OrderItem',
  id: casual.uuid,
  image: {
    image: `${casual.word}.jpg`,
  },
  name: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
})

const fakeOrder = () => ({
  __typename: 'Order',
  id: 'ord123',
  charge: 'ch_123',
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: '2022-12-11T20:16:13.797Z',
  user: fakeUser(),
})

const fakeCartItem = (overrides = {}) => ({
  __typename: 'CartItem',
  id: 'omg123',
  quantity: 3,
  product: fakeItem(),
  user: fakeUser(),
  ...overrides,
})

// Fake LocalStorage
class LocalStorageMock {
  store: any

  constructor() {
    this.store = {}
  }

  clear() {
    this.store = {}
  }

  getItem(key: string) {
    return this.store[key] || null
  }

  setItem(key: string, value: any) {
    this.store[key] = value.toString()
  }

  removeItem(key: string) {
    delete this.store[key]
  }
}

function makePaginationMocksFor(length: number) {
  return [
    {
      request: { query: PAGINATION_QUERY },
      result: {
        data: {
          _allProductsMeta: {
            count: length,
          },
          itemsConnection: {
            __typename: 'aggregate',
            aggregate: {
              count: length,
              __typename: 'count',
            },
          },
        },
      },
    },
  ]
}

export {
  makePaginationMocksFor,
  LocalStorageMock,
  fakeItem,
  fakeUser,
  fakeCartItem,
  fakeOrder,
  fakeOrderItem,
}
