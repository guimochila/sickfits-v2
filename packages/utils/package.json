{
  "name": "@sickfits/utils",
  "private": true,
  "version": "0.0.1",
  "main": "build/index.js",
  "types": "build/index.d.ts",
  "license": "MIT",
  "author": {
    "name": "Guilherme Scaldelai"
  },
  "devDependencies": {
    "@types/jest": "^28.1.0",
    "babel-jest": "^28.1.0",
    "jest": "^28.1.0",
    "ts-jest": "^28.0.4",
    "typescript": "^4.6.2"
  },
  "scripts": {
    "build": "tsc",
    "test": "NODE_ENV=test jest",
    "test:watch": "NODE_ENV=test jest --watch"
  }
}
