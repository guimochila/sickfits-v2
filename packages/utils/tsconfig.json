{
  "compilerOptions": {
    "target": "es5",
    "module": "commonjs",
    "declaration": true,
    "outDir": "./build",
    "strict": true,
    "skipLibCheck": true
  },
  "include": ["src/**/*"],
  "exclude": ["node_modules", "src/**/*.test.ts"]
}
