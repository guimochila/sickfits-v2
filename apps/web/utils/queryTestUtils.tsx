import { render } from '@testing-library/react'
import { graphql } from 'msw'
import * as React from 'react'
import { QueryClient, QueryClientProvider } from 'react-query'
import { fakeCartItem, fakeItem, fakeUser } from './testUtils'

const product = fakeItem()

export const handlers = [
  // Single Product Query mock
  graphql.query('SINGLE_PRODUCT_QUERY', (req, res, ctx) => {
    const { id } = req.variables

    if (id === product.id) {
      return res(
        ctx.data({
          product,
        }),
      )
    }

    return res(ctx.errors([{ message: 'Item not found!' }]))
  }),

  // Current User Query mock
  graphql.query('CURRENT_USER_QUERY', (req, res, ctx) => {
    const isAuthenticated = sessionStorage.getItem('is-authenticated')

    if (!isAuthenticated) {
      return res(
        ctx.data({
          authenticatedItem: fakeUser(),
        }),
      )
    }

    if (isAuthenticated === 'yesWithCartItems') {
      return res(
        ctx.data({
          authenticatedItem: fakeUser({
            cart: [fakeCartItem()],
          }),
        }),
      )
    }

    return res(
      ctx.data({
        authenticatedItem: fakeUser(),
      }),
    )
  }),

  // Pagination mock
  graphql.query('PAGINATION_QUERY', (req, res, ctx) => {
    return res(
      ctx.data({
        productsCount: 18,
      }),
    )
  }),

  // Signup Mutation mock
  graphql.mutation('SIGNUP_MUTATION', (req, res, ctx) => {
    const { name, email } = req.variables

    return res(
      ctx.data({
        createUser: {
          name,
          email,
        },
      }),
    )
  }),

  // Reset Request Mutation mock
  graphql.mutation('REQUEST_RESET_MUTATION', (req, res, ctx) => {
    const { email } = req.variables

    if (email) {
      return res(
        ctx.data({
          sendUserPasswordResetLink: true,
        }),
      )
    }

    return res(
      ctx.data({
        sendUserPasswordResetLink: null,
      }),
    )
  }),

  // Create Mutation mock
  graphql.mutation('CREATE_PRODUCT_MUTATION', (req, res, ctx) => {
    const { name, price, description } = req.variables

    if (name && price && description) {
      return res(
        ctx.data({
          createProduct: fakeItem(),
        }),
      )
    }

    return res(ctx.errors([{ message: 'Something went wrong' }]))
  }),
]

const createTestQueryClient = () =>
  new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  })

export function renderWithClient(ui: React.ReactElement) {
  const testQueryClient = createTestQueryClient()
  const { rerender, ...result } = render(
    <QueryClientProvider client={testQueryClient}>{ui}</QueryClientProvider>,
  )
  return {
    ...result,
    rerender: (rerenderUi: React.ReactElement) =>
      rerender(
        <QueryClientProvider client={testQueryClient}>
          {rerenderUi}
        </QueryClientProvider>,
      ),
  }
}

export function createWrapper() {
  const testQueryClient = createTestQueryClient()
  // eslint-disable-next-line react/display-name
  return ({ children }: { children: React.ReactNode }) => (
    <QueryClientProvider client={testQueryClient}>
      {children}
    </QueryClientProvider>
  )
}
