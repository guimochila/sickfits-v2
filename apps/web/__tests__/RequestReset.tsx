import { renderWithClient } from '../utils/queryTestUtils'
import { screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import RequestReset from '../components/RequestReset'
import { RouterContext } from 'next/dist/shared/lib/router-context'
import { createMockRouter } from '../utils/createMockRouter'

const email = 'testing@test.com'

describe('<RequestReset/>', () => {
  it('renders and matches snapshot', () => {
    const router = createMockRouter({})
    const { container } = renderWithClient(
      <RouterContext.Provider value={router}>
        <RequestReset />
      </RouterContext.Provider>,
    )

    expect(container).toMatchSnapshot()
  })

  it('calls the mutation when submitted', async () => {
    const router = createMockRouter({})
    const { container, debug } = renderWithClient(
      <RouterContext.Provider value={router}>
        <RequestReset />
      </RouterContext.Provider>,
    )

    await userEvent.type(
      screen.getByPlaceholderText('Your email address'),
      email,
    )
    await userEvent.click(screen.getByText(/Request Reset/))
    const successMessage = await screen.findByText(/Success!/i)
    expect(successMessage).toBeInTheDocument()
  })
})
