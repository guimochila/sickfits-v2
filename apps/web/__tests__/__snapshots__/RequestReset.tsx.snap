// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RequestReset/> renders and matches snapshot 1`] = `
<div>
  <form
    class="css-3eqa4t"
    method="post"
  >
    <h2>
      Request Password reset
    </h2>
    <fieldset>
      <label
        for="email"
      >
        Email
        <input
          autocomplete="email"
          name="email"
          placeholder="Your email address"
          type="email"
          value=""
        />
      </label>
      <button
        type="submit"
      >
        Request Reset
      </button>
    </fieldset>
  </form>
</div>
`;
